### PR TITLE
[Feature][Registry] Add DS K8S registry plugin

### LINKED ISSUE #13445: [Feature][Registry] Remove dependency on ZooKeeper when deployed in K8S
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

* Add a new registry plugin for DS so that DS will no longer need to depend on ZooKeeper / MySQL / ETCD for service discovery and distributed lock when deployed in K8S.
* Service Events - K8S Watch - We may use `Fabric8 Kubernetes Java Client Informer API` for better stability and less pressure on K8S api server.
* ***HeartBeat*** Instead of making services (masters / workers) persist their heartbeats, we could try to use `k8s top` to query related metrics, however, DS uses heartbeat to transmit some server runtime data besides metrics. Maybe a better way to transmit heartbeat is to put heartbeat data into pod environment variables / files. Alternatively, we could create a `configmap` for each master / worker to maintain the heartbeat data.

```java
try (InputStream is = client.pods().inNamespace(currentNamespace).withName(pod1.getMetadata().getName()).file("/msg").read())  {
  String result = new BufferedReader(new InputStreamReader(is)).lines().collect(Collectors.joining("\n"));
}
```

* Service Discovery - K8S Pod Selector / K8S Endpoints Controller
* Master Fault Tolerance (Distributed Lock) - K8S Resource Annotation - Masters will attempt to put its pod ip into the `metadata.annotations` of a resource and the one which does this successfully will be the leader to perform failover operations. - Follow-up: In fabric k8s java client, there have been implemented leader election apis using resource lock (lease, configmap, etc.). Therefore, we do not need to care too much about the details on the implementation of this distributed lock. We would make it configurable and users could choose between `lease` and `configmap` for resource lock.
* Two examples for java k8s client for leader election: 
[example with kubernetes-client](https://github.com/kubernetes-client/java/blob/master/examples/examples-release-15/src/main/java/io/kubernetes/client/examples/LeaderElectionExample.java) and 
[example with fabric8 kubernetes-client](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/LeaderElectionExamples.java)

### Design

image

### Security

* Use `Kubernetes RBAC` for security purpose. Users could control permissions on workers' access to `Kubernetes` APIs. 

### Use case

Already described above.

### Related issues

None

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by EricGao888**
## Purpose of the pull request

* WIP
* This PR will close: #13445

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by EricGao888** [comment]
Action Items:

1. Use `Informer` instead of plain `Watch`.
2. Restrict K8S client permissions with K8S RBAC for masters and workers.

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import lombok.Data;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+@ConditionalOnProperty(prefix = "registry", name = "type", havingValue = "k8s")
```
Comment (kezhenxu94) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java:30]: I prefer we use the complete form

```suggestion
@ConditionalOnProperty(prefix = "registry", name = "type", havingValue = "kubernetes")
```

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+@ConfigurationProperties(prefix = "registry")
+public class KubernetesProperties {
+
+    private String endpoints;
+    private String namespace = "dolphinscheduler";
+    private Duration connectionTimeout = Duration.ofSeconds(9);
+
+    // TODO: add rbac related config
```
Comment (kezhenxu94) [inline dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java:38]: Hi, I don't think we need this, I believe `kubernetes` registry plugin must be used when DolphinScheduler itself is deployed inside kubernetes, so we should add the rbac in our helm chart and we can assume it has  sufficient permissions to access the needed resources, so no extra configurations are needed here

File: dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java (hunk 1 above)
Comment (EricGao888) [reply dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-kubernetes/src/main/java/org/apache/dolphinscheduler/plugin/registry/kubernetes/KubernetesProperties.java:30]: Sure, I will change it.

**Pull request closed without merging**